### PR TITLE
method triangulate

### PR DESCRIPTION
**Pull request by okrdima**
Finish thriangulate and simplify methods

### CODE REVIEW

**Comment by okrdima** [comment]
Also i develop triangulate method with centroid and i testing him
image

File: GeometryCore/src/Polygon.cpp [hunk 1]
```diff
@@ -177,9 +177,31 @@ namespace geom_utils
     // For non-convex polygons it should return just empty vector.
     std::vector<Triangle2D> Polygon::triangulate() const
     {
-        return {}; 
+        std::vector <Triangle2D> vectorForTriangulate;
+
+       // FPoint2D centerPoly = centroid();
```
Comment (VladGribinchuk) [inline GeometryCore/src/Polygon.cpp]: Never commit commented out code lines (at least, never let them get into pull requests)
So, please clean all of those

File: GeometryCore/src/Polygon.cpp [hunk 2]
```diff
@@ -177,9 +177,31 @@ namespace geom_utils
...
+       // FPoint2D  firstVertex =  pointsCopy[0];
+        if (points.size() >= 4) {
+            if (isConvexHull())
+            {
+              //  auto vertexSortingCondition = [](const FPoint2D& a, const FPoint2D& b) {return a.x < b.x || (a.x == b.x && a.y < b.y); };
+                auto pointsCopy = this->points;
```
Comment (VladGribinchuk) [inline GeometryCore/src/Polygon.cpp]: Do you really need copy of points? Or we can use just points?

File: GeometryCore/src/Polygon.cpp [hunk 3]
```diff
@@ -177,9 +177,31 @@ namespace geom_utils
...
+                //std::sort(pointsCopy.begin(), pointsCopy.end(), vertexSortingCondition);
+                FPoint2D  firstVertex =  pointsCopy[0];
+                for (int i = 0; i < pointsCopy.size() - 2 ; i++)
+                {
+                   
+                    Triangle2D tri = { pointsCopy[(i + 1) % pointsCopy.size()] ,firstVertex,pointsCopy[(i+2) % pointsCopy.size()] };
```
Comment (VladGribinchuk) [inline GeometryCore/src/Polygon.cpp]: The output triangles should have the same orientation like the polygon. Lets see what we have here.
image

Please fix

File: GeometryCore/src/Polygon.cpp [hunk 4]
```diff
@@ -177,9 +177,31 @@ namespace geom_utils
...
+                    vectorForTriangulate.push_back(tri);
+                }
+                return vectorForTriangulate;
+            }
+            else return vectorForTriangulate;
```
Comment (VladGribinchuk) [inline GeometryCore/src/Polygon.cpp]:
```
    return vectorForTriangulate;
            }
            else return vectorForTriangulate;
```

verbose: you can use just one `return vectorForTriangulate;` statement

File: UnitTests/PolygonTests.h [hunk 5]
```diff
@@ -151,4 +152,44 @@ DEFINE_TEST_CASE(PolygonTranslate)
...
+
+
+DEFINE_TEST_CASE(PolygonTriangulation)
+{
+    Polygon poly ({ {-3,-4},{-4, 0}, {0, 4}, {4,0}, {3,-4} });
+
+
+    // {-4,0}, {-3,-4},{0, 4}, {4, 0}, {3,-4}
+
+    /*std::vector <Triangle2D> vectorForTriangulate = {
```
Comment (VladGribinchuk) [inline UnitTests/PolygonTests.h]: Do not commit commented out code. Leave them only locally during your work.

**Review by VladGribinchuk — CHANGES_REQUESTED** [review]

**Review by VladGribinchuk — COMMENTED** [review]
Looks good now 👌 
Please continue your work with simplification method

**Review by VladGribinchuk — APPROVED** [review]

File: GeometryCore/src/Polygon.cpp [hunk 6]
```diff
@@ -177,9 +177,27 @@ namespace geom_utils
...
+           
+                FPoint2D  firstVertex =  points[0];
+                for (int i = 0; i < points.size() - 2 ; i++)
+                {
+                   
+                    Triangle2D tri = {points[(i + 2) % points.size()] ,firstVertex, points[(i+1) % points.size()] };
+                    vectorForTriangulate.push_back(tri);
+                }
+            }
+            return vectorForTriangulate;
```
Comment (VladGribinchuk) [inline GeometryCore/src/Polygon.cpp]: What method should return if points.size() < 4?

**Review by VladGribinchuk — CHANGES_REQUESTED** [review]
sorry, one more problem

File: GeometryCore/src/Polygon.cpp (hunk 6 above)
Comment (okrdima) [reply GeometryCore/src/Polygon.cpp]: Does he also have to return an empty vector?

Comment (VladGribinchuk) [reply GeometryCore/src/Polygon.cpp]: Lets think. In general, we can have the following input:
- empty polygon or 1- or 2- points polygon
- 3 or more points polygon

Please think what output should be in those cases.

Please analyse carefully **input** and **output** each time while developing some method/function etc.

**Comment by VladGribinchuk** [comment]
@IlliaHryshchuk and @1m8o5s 
Please have a review of this PR. Check each method, each code line. Check the whole picture. Approve these changes or comment any of your concerning. 
I will NOT merge this without your feedbacks.

File: GeometryCore/src/Polygon.cpp [hunk 7]
```diff
@@ -120,10 +121,92 @@ namespace geom_utils
     }
 
 
+    float Polygon::perpendicularDistance(const FPoint2D& pointsList,const FPoint2D& lineStart, const FPoint2D& lineEnd)
+    {
+        float dx = lineEnd.x - lineStart.x;
+        float dy = lineEnd.y - lineStart.y;
```
Comment (IlliaHryshchuk) [inline GeometryCore/src/Polygon.cpp]: - Why don't you use just `FPoint2D someName(lineEnd.x - lineStart.x, lineEnd.y - lineStart.y);`? we have an already written struct for points
 - also we have already written distance method in `LinearAlgebra.h`

File: GeometryCore/src/Polygon.cpp [hunk 8]
```diff
@@ -120,10 +121,92 @@ namespace geom_utils
...
+        {
+            dx /= mag; 
+            dy /= mag;
+        }
+
+        float pvx = pointsList.x - lineStart.x;
+        float pvy = pointsList.y - lineStart.y;
+
+        //Get dot product (project pv onto normalized direction)
+        float pvdot = dx * pvx + dy * pvy;
```
Comment (IlliaHryshchuk) [inline GeometryCore/src/Polygon.cpp]: We have already written dot product. Please, check the `LinearAlgebra.h`

File: GeometryCore/src/Polygon.cpp [hunk 9]
```diff
@@ -120,10 +121,92 @@ namespace geom_utils
     }
 
 
+    float Polygon::perpendicularDistance(const FPoint2D& pointsList,const FPoint2D& lineStart, const FPoint2D& lineEnd)
```
Comment (IlliaHryshchuk) [inline GeometryCore/src/Polygon.cpp]: - why a single point names `pointsList` if It is not a list?

File: GeometryCore/src/Polygon.cpp [hunk 10]
```diff
@@ -120,10 +121,92 @@ namespace geom_utils
...
+        float dsy = pvdot * dy;
+
+        //Subtract this from pointsList
+        float ax = pvx - dsx;
+        float ay = pvy - dsy;
+
+        return pow(pow(ax, 2.0) + pow(ay, 2.0), 0.5);
+    
+    }
+    void Polygon::recursiveAlgorithmForSimplify(const std::vector<FPoint2D>& pointList, const FPoint2D::coord smallestLineLength, std::vector<FPoint2D>& out)
```
Comment (IlliaHryshchuk) [inline GeometryCore/src/Polygon.cpp]: - as for me, the name of function is very complicated

**Review by IlliaHryshchuk — CHANGES_REQUESTED** [review]
To my mind, the `Polygon::perpendicularDistance` could be improved

**Comment by VladGribinchuk** [comment]
@1m8o5s Nikolay what do you think? Please let us know your review when you will have a chance to take a look.

File: GeometryCore/src/Polygon.cpp [hunk 11]
```diff
@@ -120,10 +121,88 @@ namespace geom_utils
...
+        FPoint2D pv(point-lineStart);
+
+        //Get dot product (project pv onto normalized direction)
+        float pvdot = dot(dpoint,pv);
+       
+        //Scale line direction vector
+        FPoint2D ds(pvdot * dpoint.x, pvdot * dpoint.y);
+
+        //Subtract this from pointv
+        FPoint2D ap(pv.x - ds.x, pv.y - ds.y);
```
Comment (1m8o5s) [inline GeometryCore/src/Polygon.cpp]: We have realization of projection point to line segment in LinearAlgebra

**Comment by okrdima** [comment]
As approximately does the algorithm
image

File: GeometryCore/src/Polygon.cpp (hunk 11 above)
Comment (okrdima) [reply GeometryCore/src/Polygon.cpp]: Yes, i see, good idea, I'll fix it now

**Comment by VladGribinchuk** [comment]
@IlliaHryshchuk @1m8o5s @okrdima Good job guys! A few more thoughts from my side in conclusion. 

1. `void algorithmSimplify` - this name looks verbose. You can call it just `simplify` (overloaded version of current `simplify`). Or you can call it even `simplifyRamerDouglasPeucker()` - thus, explicitly show the algorithm used for simplification. 
Also should it be public, what do you think? Encapsulation. It looks like smth internal rather than smth exposed for the user of Polygon class.
2. https://github.com/VladGribinchuk/FoundationBuilder/pull/43/files#diff-a6823a64239e0f65b606432fc8fc6fe9L123-L124
```
    // Remove line segments which are smaller than the provided smallestLineLength value.
    // Result will recall smth like re-connecting points process.
```
Since you decided to implement different algorithm for simplification, this description is outdated and will confuse user. Please provide new description for algorithm + link to wiki or page where user can find this algorithm. Also, new algorithm doesn't use `smallestLineLength`, so this argument also should be renamed to be more appropriate.
3. https://github.com/VladGribinchuk/FoundationBuilder/pull/43/files#diff-a6823a64239e0f65b606432fc8fc6fe9R233
`if (points.size() > 2) {`
this check looks redundant.

**Comment by okrdima** [comment]
@VladGribinchuk  Why
 if (points.size() > 2)
this check looks redundant? If the size is less than two, you can't simplify the polygon

**Comment by VladGribinchuk** [comment]
> @VladGribinchuk Why
> if (points.size() > 2)
> this check looks redundant? If the size is less than two, you can't simplify the polygon

https://github.com/VladGribinchuk/FoundationBuilder/pull/43/files#diff-a6823a64239e0f65b606432fc8fc6fe9R233
Triangulate function

**Comment by okrdima** [comment]
@VladGribinchuk How to build Triangle with 2 Point?

**Comment by VladGribinchuk** [comment]
> @VladGribinchuk How to build Triangle with 2 Point?

In no way) 
Please have a look at the loop you implemented in triangulate() `for (int i = 0; i < points.size() - 2 ; i++)`. Assume there are 2 or less points. How many iterations will be carried on? Then think if you really need that additional check `if (points.size() > 2)`.

**Comment by okrdima** [comment]
@VladGribinchuk Ok,i agree with you

**Review by VladGribinchuk — APPROVED** [review]

**Pull request merged**